pallets/gear-builtin/src/staking.rs

```diff
@@ -0,0 +1,178 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Staking builtin actor implementation
+
+use super::*;
+use common::Origin;
+use core::marker::PhantomData;
+use frame_support::dispatch::{extract_actual_weight, GetDispatchInfo};
+use gbuiltin_staking::*;
+use pallet_staking::RewardDestination;
+use parity_scale_codec::Decode;
+use sp_runtime::traits::{Dispatchable, StaticLookup, UniqueSaturatedInto};
+
+pub struct Actor<T: Config>(PhantomData<T>);
+
+impl<T: Config> Actor<T>
+where
+    T::AccountId: Origin,
+{
+    fn dispatch_call(
+        origin: T::AccountId,
+        call: <T as Config>::RuntimeCall,
+        gas_limit: u64,
+    ) -> (Result<(), BuiltinActorError>, u64) {
+        let call_info = call.get_dispatch_info();
+
+        // Necessary upfront gas sufficiency check
+        if gas_limit < call_info.weight.ref_time() {
+            return (Err(BuiltinActorError::InsufficientGas), 0_u64);
+        }
+
+        // Execute call
+        let res = call.dispatch(frame_system::RawOrigin::Signed(origin).into());
+        let actual_gas = extract_actual_weight(&res, &call_info).ref_time();
+        match res {
+            Ok(_post_info) => {
+                log::debug!(
+                    target: LOG_TARGET,
+                    "Call dispatched successfully",
+                );
+                (Ok(()), actual_gas)
+            }
+            Err(e) => {
+                log::debug!(target: LOG_TARGET, "Error dispatching call: {:?}", e);
+                (
+                    Err(BuiltinActorError::Custom(LimitedStr::from_small_str(
+                        e.into(),
+                    ))),
+                    actual_gas,
+                )
+            }
+        }
+    }
+}
+
+impl<T: Config> BuiltinActor for Actor<T>
+where
+    T::AccountId: Origin,
+{
+    const ID: u64 = 2;
+
+    type Error = BuiltinActorError;
+
+    fn handle(dispatch: &StoredDispatch, gas_limit: u64) -> (Result<Payload, Self::Error>, u64) {
+        let message = dispatch.message();
+        let origin: T::AccountId = dispatch.source().cast();
+        let mut payload = message.payload_bytes();
+
+        let decoding_cost =
```

Comment:
shouldn't gas limit be decreased by decoding cost here?

Reply:
It should and it is (just a couple of lines further down before deriving a RuntimeCall).